src/rgw/driver/posix/rgw_sal_posix.h

```diff
@@ -0,0 +1,718 @@
+
+  virtual int list_buckets(const DoutPrefixProvider* dpp,
+			   const std::string& marker, const std::string& end_marker,
+			   uint64_t max, bool need_stats, BucketList& buckets,
+			   optional_yield y) override;
+  virtual int create_bucket(const DoutPrefixProvider* dpp,
+                            const rgw_bucket& b,
+                            const std::string& zonegroup_id,
+                            rgw_placement_rule& placement_rule,
+                            std::string& swift_ver_location,
+                            const RGWQuotaInfo* pquota_info,
+                            const RGWAccessControlPolicy& policy,
+                            Attrs& attrs,
+                            RGWBucketInfo& info,
+                            obj_version& ep_objv,
+                            bool exclusive,
+                            bool obj_lock_enabled,
+                            bool* existed,
+                            req_info& req_info,
+                            std::unique_ptr<Bucket>* bucket,
+                            optional_yield y) override;
+  virtual Attrs& get_attrs() override { return next->get_attrs(); }
+  virtual void set_attrs(Attrs& _attrs) override { next->set_attrs(_attrs); }
+  virtual int read_attrs(const DoutPrefixProvider* dpp, optional_yield y) override;
+  virtual int merge_and_store_attrs(const DoutPrefixProvider* dpp, Attrs&
+				    new_attrs, optional_yield y) override;
+  virtual int load_user(const DoutPrefixProvider* dpp, optional_yield y) override;
+  virtual int store_user(const DoutPrefixProvider* dpp, optional_yield y, bool
+			 exclusive, RGWUserInfo* old_info = nullptr) override;
+  virtual int remove_user(const DoutPrefixProvider* dpp, optional_yield y) override;
+};
+
+class POSIXBucket : public StoreBucket {
+private:
+  POSIXDriver* driver;
+  int parent_fd{-1};
+  int dir_fd{-1};
+  struct statx stx;
+  bool stat_done{false};
+  RGWAccessControlPolicy acls;
+  std::optional<std::string> ns;
+
+public:
+  POSIXBucket(POSIXDriver *_dr, int _p_fd, const rgw_bucket& _b, User* _u, std::optional<std::string> _ns = std::nullopt)
+    : StoreBucket(_b, _u),
+    driver(_dr),
+    parent_fd(_p_fd),
+    acls(),
+    ns(_ns)
+    { }
+
+  POSIXBucket(POSIXDriver *_dr, int _p_fd, const RGWBucketEnt& _e, User* _u)
+    : StoreBucket(_e, _u),
+    driver(_dr),
+    parent_fd(_p_fd),
+    acls()
+    { }
+
+  POSIXBucket(POSIXDriver *_dr, int _p_fd, const RGWBucketInfo& _i, User* _u)
+    : StoreBucket(_i, _u),
+    driver(_dr),
+    parent_fd(_p_fd),
+    acls()
+    { }
+
+  POSIXBucket(POSIXBucket& _o) :
+    StoreBucket(_o),
+    driver(_o.driver),
+    parent_fd(_o.parent_fd),
+    /* Don't want to copy dir_fd */
+    stx(_o.stx),
+    stat_done(_o.stat_done),
+    acls(_o.acls),
+    ns(_o.ns) {}
+
+  virtual ~POSIXBucket() { close(); }
+
+  virtual void set_owner(rgw::sal::User* _owner) override {
+    StoreBucket::set_owner(_owner);
+    info.owner = owner->get_id();
```

Comment:
isn't `info` a member of `StoreBucket`? why doesn't its `set_owner()` handle this?

Reply:
good question